### PR TITLE
Add vmap support for `smooth_l1_loss_backward`

### PR DESCRIPTION
**Pull request by yhl48**
Follow-up of #98357

### CODE REVIEW

**Comment by kshitij12345** [comment]
You will have to update the following file (i.e. remove smooth_l1_loss_backward) to get the test passing. The failing test compares whether the ops without decomposition match ops listed in this file.

https://github.com/pytorch/pytorch/blob/main/test/expect/HasDecompTest.test_has_decomposition.expect

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -362,6 +362,21 @@ def mse_loss_backward(
...
+    grad_output: Tensor, self: Tensor, target: Tensor, reduction: int, beta: float
+):
+    norm = 1.0 / self.numel() if reduction == Reduction.MEAN.value else 1.0
+    x = self - target
+    abs_x = torch.abs(x)
+    return torch.where(
+        abs_x < beta,
+        norm * grad_output * x / beta,
+        norm * grad_output * torch.sign(x),
+    )
```
[the remark this answers is not in the document]
Comment (yhl48) [reply torch/_decomp/decompositions.py:377]: Fixed in https://github.com/pytorch/pytorch/pull/99429/commits/55554ae96a3ed6104e51edb68193f0fc89bafbf8, thanks!

**Comment by yhl48** [comment]
Locally, the following test fails

```
functorch/test_ops.py::TestOperatorsCPU::test_vmapvjpvjp_linalg_tensorsolve_cpu_float32 - AssertionError: Tensor-likes are not close
```

But there does not seem to be an error in the Github workflow, would this be a cause for concern?

**Comment by zou3519** [comment]
@yhl48  what does the complete error message say?

**Comment by yhl48** [comment]
@zou3519 

```
Traceback (most recent call last):
  File "/pytorch/test/functorch/test_ops.py", line 862, in test_vmapvjpvjp
    self.assertEqual(loop_out, batched_out)
  File "/pytorch/torch/testing/_internal/common_utils.py", line 3031, in assertEqual
    raise error_metas[0].to_error(
AssertionError: Tensor-likes are not close!

Mismatched elements: 2 / 288 (0.7%)
Greatest absolute difference: 0.0002994537353515625 at index (0, 1, 0, 0, 2) (up to 0.0001 allowed)
Greatest relative difference: 0.0001621410046936944 at index (0, 1, 0, 0, 2) (up to 0.0001 allowed)
```

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -362,6 +362,36 @@ def mse_loss_backward(
...
+@pw_cast_for_opmath
+def smooth_l1_loss_backward(
+    grad_output: Tensor, self: Tensor, target: Tensor, reduction: int, beta: float
+):
+    norm = 1.0 / self.numel() if reduction == Reduction.MEAN.value else 1.0
+    x = self - target
+    abs_x = torch.abs(x)
+    return torch.where(
+        abs_x < beta,
+        norm * grad_output * x / beta,
```
Comment (kshitij12345) [inline torch/_decomp/decompositions.py:375]: We should save `norm * grad_output` to a variable to avoid computing it twice.

**Comment by kshitij12345** [comment]
> Mismatched elements: 2 / 288 (0.7%)
Greatest absolute difference: 0.0002994537353515625 at index (0, 1, 0, 0, 2) (up to 0.0001 allowed)
Greatest relative difference: 0.0001621410046936944 at index (0, 1, 0, 0, 2) (up to 0.0001 allowed)

Difference is close to allowed difference. I think this is ok. Wdyt @zou3519 ?

**Review by kshitij12345 — APPROVED** [review]
LGTM! Thank you @yhl48

**Comment by kshitij12345** [comment]
@zou3519 will have to accept this PR (as it touches part of PyTorch besides functorch). Thanks for the reminder.

Ping @zou3519.

**Review by kshitij12345 — CHANGES_REQUESTED** [review]
Also, it would be great to add similar comment. Thanks!

https://github.com/pytorch/pytorch/blob/cc01568efd0037ce201efecccb165862655f4da9/aten/src/ATen/functorch/BatchRulesLoss.cpp#L315

**Review by kshitij12345 — APPROVED** [review]
LGTM

**Review by zou3519 — APPROVED** [review]

**Comment by kshitij12345** [comment]
ROCM failure looks irrelevant

```
023-04-27T20:44:19.8389536Z =================================== FAILURES ===================================
2023-04-27T20:44:19.8389905Z ___________________________ TestProfiler.test_kineto ___________________________
2023-04-27T20:44:19.8390265Z Traceback (most recent call last):
2023-04-27T20:44:19.8390645Z   File "profiler/test_profiler.py", line 747, in test_kineto
2023-04-27T20:44:19.8391014Z     self.assertTrue(found_gemm)
2023-04-27T20:44:19.8391434Z   File "/opt/conda/envs/py_3.8/lib/python3.8/unittest/case.py", line 765, in assertTrue
2023-04-27T20:44:19.8391852Z     raise self.failureException(msg)
2023-04-27T20:44:19.8392199Z AssertionError: False is not true
2023-04-27T20:44:19.8392950Z - generated xml file: /var/lib/jenkins/pytorch/test/test-reports/python-pytest/profiler.test_profiler/profiler.test_profiler-3e6bfcbb35fbe445.xml -
2023-04-27T20:44:19.8393555Z =========================== short test summary info ============================
2023-04-27T20:44:19.8394153Z FAILED [0.0097s] profiler/test_profiler.py::TestProfiler::test_kineto - AssertionError: False is not true
2023-04-27T20:44:19.8394640Z !!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
2023-04-27T20:44:19.8395056Z !!!!!!!!!!!!!!! Test failed, continuing from this test next run. !!!!!!!!!!!!!!!
2023-04-27T20:44:19.8395460Z ==================== 1 failed, 15 passed, 2 rerun in 4.53s =====================
2023-04-27T20:44:19.8396167Z FINISHED PRINTING LOG FILE of profiler/test_profiler 1/1 (/var/lib/jenkins/pytorch/test/test-reports/profiler-test_profiler_lt4dskct.log)
```

@pytorchbot merge -f

File: torch/_decomp/decompositions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/_decomp/decompositions.py:378]: This decomposition is subject to the bug reported here #96813 when beta = 0 (abs_x can be 0, `self-target` is computed potentially imprecisely

**Pull request closed without merging**